spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/MaterializingIterator.scala

```diff
@@ -0,0 +1,37 @@
+package com.datastax.spark.connector.util
+
+import java.util.NoSuchElementException
+
+import scala.collection.mutable
+
+/** Materializes/Fetches up to `materializationLimit` from underlying `iterator`.
+  * Materialization happens on `next()` invocation.
+  *
+  * In following example `it.next()` is invoked three times, results are caches in internal queue.
+  * {{{
+  * new MaterializingIterator(it, 3).next()
+  * }}}
+  *
+  */
+class MaterializingIterator[T](iterator: Iterator[T], materializationLimit: Int) extends Iterator[T] {
+
+  assert(materializationLimit > 0, "Materialization limit must be greater than 0")
+
+  private val materialized: mutable.Queue[T] = new mutable.Queue[T]()
+
+  private def materialize(): Unit = {
+    while (iterator.hasNext && materialized.size < materializationLimit) {
+      materialized.enqueue(iterator.next)
+    }
+  }
+
+  override def hasNext = !materialized.isEmpty || iterator.hasNext
+
+  override def next() = {
+    if(!hasNext) {
+      throw new NoSuchElementException("No more elements")
+    }
+    materialize()
```

Comment:
Could be. I think the lazy loading is a probably a benefit but i'm not sure how much it really helps. If the next operation is a shuffle the whole block will have to be materialized anyway.

I'm mostly worried though about using a straight queue here, If we have Queries [a,b,c,d] In my queue, and a takes say 1 minute to pull, but b,c,d all take 1ms. I'm going to wait 1 minute without issuing any new queries.